### PR TITLE
Elementwise mutliplications

### PR DESCRIPTION
**Pull request by chillenzer**
This PR adds elementwise multiplication spectra in real and spectral space. The real space version is very subtle. I think I've got it working for quadrature lattices but I'm not done thinking about more general lattices. This is another step towards a physical application. Combined with #50, #51 and #53  this will enable us to write down an effective action.

### CODE REVIEW

**Review by Mayank447 — COMMENTED** [review]
General suggestion please specify the basis when using apply_to method of the Dirac Operator. Also one question why did you change to np.inner, just curious

File: pseudospectral/spectra/element_wise.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+class ElementwiseSpectralMultiplication:
+    """
+    This spectrum represents elementwise multiplication of the spectral coefficients.
+    In a way, it generalises the specialised spectra. For example, using
+    spectrum.eigenvalues as field_values reproduces the original spectrum.
+    """
+
+    def __new__(cls, spectrum, given_field_values=1):
+        class TailoredElementwiseSpectralMultiplication(
+            type(spectrum), ElementwiseSpectralMultiplication
```
Comment (Mayank447) [inline pseudospectral/spectra/element_wise.py:15]: Add a comment explaining it (isinstance)

File: pseudospectral/spectra/element_wise.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+            # Revert the non-trivial effects of inheriting from
+            # ElementwiseRealMultiplication.
+            def __new__(cls, *args, **kwargs):
+                return super().__new__(cls, *args, **kwargs)
+
+            # We'll not initialise it this way and need to avoid wrong constructor
+            # calls.
+            def __init__(self, *args, **kwargs):
+                pass
```
Comment (Mayank447) [inline pseudospectral/spectra/element_wise.py:82]: Recheck whether this is required

File: pseudospectral/spectra/element_wise.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+                else:
+                    message = (
+                        "Unsupported space transformation from "
+                        f"{input_basis} to {output_basis}."
+                    )
+                    raise ValueError(message)
+
+            def scalar_product(self, lhs, rhs, input_basis="real"):
+                return super().scalar_product(lhs, rhs, input_basis)
```
Comment (Mayank447) [inline pseudospectral/spectra/element_wise.py:124]: Remove these lines

File: pseudospectral/spectra/element_wise.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        )
+        norms = np.sqrt(
+            np.diag(
+                spectrum.scalar_product(
+                    spectral_coefficients,
+                    spectral_coefficients,
+                    "spectral",
+                )
+            )
+        )
```
Comment (Mayank447) [inline pseudospectral/spectra/element_wise.py:62]: Improve the complexity, currently we are computing N^2 scalar products

File: pseudospectral/spectra/element_wise.py (hunk 1 above)
Comment (chillenzer) [reply pseudospectral/spectra/element_wise.py:19]: Done.

File: pseudospectral/spectra/element_wise.py (hunk 2 above)
Comment (chillenzer) [reply pseudospectral/spectra/element_wise.py:82]: Done.

File: pseudospectral/spectra/element_wise.py (hunk 3 above)
Comment (chillenzer) [reply pseudospectral/spectra/element_wise.py]: Done.

File: pseudospectral/spectra/element_wise.py (hunk 4 above)
Comment (chillenzer) [reply pseudospectral/spectra/element_wise.py:63]: I've opened #57 to track this. There are multiple ways to go from here and I think that we should postpone this to a separate PR.

**Comment by chillenzer** [comment]
Think I handled everything. Rebased to resolve the conflicts. The complexity comment is delegated to #57. Please review again and merge.

**Pull request merged**